src/embed/ts-embed.ts

```diff
@@ -592,12 +627,161 @@ export class TsEmbed {
                     uploadMixpanelEvent(MIXPANEL_EVENT.VISUAL_SDK_RENDER_FAILED, {
                         error: JSON.stringify(error),
                     });
-                    this.insertIntoDOM(this.embedConfig.loginFailedMessage);
+                    if (this.isPreRendered) {
+                        this.insertIntoDOMForPreRender(this.embedConfig.loginFailedMessage);
+                    } else {
+                        this.insertIntoDOM(this.embedConfig.loginFailedMessage);
+                    }
                     this.handleError(error);
                 });
         });
     }
 
+    public getPreRenderIds() {
+        return {
+            wrapper: `tsEmbed-pre-render-wrapper-${this.viewConfig.preRenderId}`,
+            shield: `tsEmbed-pre-render-shield-${this.viewConfig.preRenderId}`,
+            child: `tsEmbed-pre-render-child-${this.viewConfig.preRenderId}`,
+        };
+    }
+
+    protected createPreRenderWrapper(child: HTMLElement): HTMLDivElement {
+        if (!this.viewConfig.preRenderId) {
+            throw new Error('Pre render id is required');
+        }
+
+        const preRenderIds = this.getPreRenderIds();
+
+        const stalePreRenderWrapper = document.getElementById(preRenderIds.wrapper);
+        if (stalePreRenderWrapper) {
+            console.log('Found stale wrapper , removing');
+            stalePreRenderWrapper.remove();
+        }
+
+        const preRenderWrapper = document.createElement('div');
+        preRenderWrapper.id = preRenderIds.wrapper;
+        setStyleProperties(preRenderWrapper, { position: 'absolute', width: '100vw', height: '100vh' });
+
+        const preRenderShield = document.createElement('div');
+        preRenderShield.id = preRenderIds.shield;
+        setStyleProperties(preRenderShield, { position: 'absolute', width: '100%', height: '100%' });
```

Comment:
Abstract away this frequently used logic.
```suggestion
const createUnqiueDivWithStyle = (id, styles) => {
    const div = document.createElement('div');
    div.setAttribute('id', id);
    setStyleProperties(div, styles);
    return div;
};

this.preRenderWrapper = createUnqiueDivWithStyle(preRenderIds.wrapper, { position: 'absolute', width: '100vw', height: '100vh' });
this.preRenderWrapper = createUnqiueDivWithStyle(preRenderIds.shield, { position: 'absolute', width: '100%', height: '100%' });
```